africanus/calibration/phase_only.py

```diff
@@ -0,0 +1,222 @@
+        def jacobian(a1j, blj, a2j, sign, out):
+            out[...] = sign * a1j * blj * np.conj(a2j)
+    elif mode == DIAG:
+        def jacobian(a1j, blj, a2j, sign, out):
+            out[...] = 0
+    elif mode == FULL:
+        def jacobian(a1j, blj, a2j, sign, out):
+            out[...] = 0
+    return njit(nogil=True)(jacobian)
+
+
+@generated_jit(nopython=True, nogil=True, cache=True)
+def jhj_and_jhr(time_indices, antenna1, antenna2, counts,
+                jones, residual, model, flag):
+
+    mode = check_type(jones, residual)
+
+    if mode:
+        raise NotImplementedError("Only DIAG-DIAG case has been implemented")
+
+    jacobian = jacobian_factory(mode)
+
+    @wraps(jhj_and_jhr)
+    def _jhj_and_jhr_fn(time_indices, antenna1, antenna2, counts,
+                        jones, residual, model, flag):
+        jones_shape = np.shape(jones)
+        tmp_out_array = np.zeros_like(jones[0, 0, 0, 0], dtype=jones.dtype)
+        n_tim = jones_shape[0]
+        n_ant = jones_shape[1]
+        n_chan = jones_shape[2]
+        n_dir = jones_shape[3]
+
+        jhr = np.zeros(jones.shape, dtype=jones.dtype)
+        jhj = np.zeros(jones.shape, dtype=jones.real.dtype)
+
+        for t in range(n_tim):
+            ind = np.arange(time_indices[t], time_indices[t] + counts[t])
+            for ant in range(n_ant):
+                # find where either antenna == ant
+                # these will be mutually exclusive since no autocorrs
+                for row in ind:
+                    if antenna1[row] == ant or antenna2[row] == ant:
+                        p = antenna1[row]
+                        q = antenna2[row]
+                        if ant == p:
+                            sign = 1.0j
+                        elif ant == q:
+                            sign = -1.0j
+                        else:
+                            raise RuntimeError(
+                                "Got impossible antenna number. This is a bug")
+                        for nu in range(n_chan):
+                            if not np.any(flag[row, nu]):
+                                for s in range(n_dir):
+                                    jacobian(
+                                        jones[t, p, nu, s],
+                                        model[row, nu, s],
+                                        jones[t, q, nu, s],
+                                        sign,
+                                        tmp_out_array)
+                                    jhj[t, ant, nu,
+                                        s] += (np.conj(tmp_out_array) *
+                                               tmp_out_array).real
+                                    jhr[t, ant, nu,
+                                        s] += (np.conj(tmp_out_array) *
+                                               residual[row, nu])
+        return jhj, jhr
+    return _jhj_and_jhr_fn
+
+
+def phase_only_GN(time_indices, antenna1, antenna2,
```

Comment:
What does GN stand for? Gains? Gauss-Newton? Expanding the acronym in the function name might make it immediately understandable.

Reply:
Changed it to phase_only_gauss_newton